### DISCUSSION TITLE
using a GitHub Packages hosted repo in a Java/maven action

### DISCUSSION THREAD

**Discussion by ghost** [community/community#25160] [Actions]
<p>Hello there,</p>
<p>I tried to look up the search function and the available sections, but couldn’t find a better spot. If there is one please feel free to point me to it.</p>
<p>Topic: I want to develop a client in Java for the <a href="http://mixer.com" rel="nofollow noopener">mixer.com</a> API. mixer has their package hosted on a GitHub Packages repo instead of pushing it to maven central. After a bit of tinkering I got it to work in NetBeans 11.2 by using the settings.xml in the ~/.m2/ directory so I’m able to build it local on my machine. As I also want to use the simple maven flow I struggle to get it to work. My current setup looks like this (as found in some other topic here):</p>
<pre><code># This workflow will build a Java project with Maven
# For more information see: https://help.github.com/actions/language-and-framework-guides/building-and-testing-java-with-maven

name: Java CI with Maven

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up JDK 13
      uses: actions/setup-java@v1
      with:
        java-version: 13
    - name: Build with Maven
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      run: mvn -B package --file pom.xml

&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"&gt;
	&lt;modelVersion&gt;4.0.0&lt;/modelVersion&gt;
	&lt;groupId&gt;de.cryptearth&lt;/groupId&gt;
	&lt;artifactId&gt;mixerclient&lt;/artifactId&gt;
	&lt;version&gt;1.0&lt;/version&gt;
	&lt;packaging&gt;jar&lt;/packaging&gt;
	&lt;properties&gt;
		&lt;project.build.sourceEncoding&gt;UTF-8&lt;/project.build.sourceEncoding&gt;
		&lt;maven.compiler.source&gt;13&lt;/maven.compiler.source&gt;
		&lt;maven.compiler.target&gt;13&lt;/maven.compiler.target&gt;
	&lt;/properties&gt;
	&lt;repositories&gt;
		&lt;repository&gt;
			&lt;id&gt;central&lt;/id&gt;
			&lt;name&gt;central&lt;/name&gt;
			&lt;url&gt;https://repo.maven.apache.org/maven2/&lt;/url&gt;
		&lt;/repository&gt;
		&lt;repository&gt;
			&lt;id&gt;mixer&lt;/id&gt;
			&lt;name&gt;mixer&lt;/name&gt;
			&lt;url&gt;https://maven.pkg.github.com/mixer/beam-client-java&lt;/url&gt;
		&lt;/repository&gt;
	&lt;/repositories&gt;
	&lt;dependencies&gt;
		&lt;dependency&gt;
			&lt;groupId&gt;com.mixer&lt;/groupId&gt;
			&lt;artifactId&gt;api&lt;/artifactId&gt;
			&lt;version&gt;6.0.0&lt;/version&gt;
		&lt;/dependency&gt;
	&lt;/dependencies&gt;
&lt;/project&gt;
</code></pre>
<p>I also had already a few hours staring at the documentation, but I just can’t get it to work. The codeflow always ends up with this issue:</p>
<pre><code>[ERROR] Failed to execute goal on project mixerclient: Could not resolve dependencies for project de.cryptearth:mixerclient:jar:1.0: Failed to collect dependencies at com.mixer:api:jar:6.0.0: Failed to read artifact descriptor for com.mixer:api:jar:6.0.0: Could not transfer artifact com.mixer:api:pom:6.0.0 from/to mixer (https://maven.pkg.github.com/mixer/beam-client-java): Authentication failed for https://maven.pkg.github.com/mixer/beam-client-java/com/mixer/api/6.0.0/api-6.0.0.pom 401 Unauthorized -&gt; [Help 1]
</code></pre>
<p>So, my simple question comes down to this: What information do I have to put where to make the simple maven codeflow work correctly to access a repo hosted on GitHub Packages using this special GITHUB-TOKEN thing?</p>
<p>I already put several hours just into this to get it working - but still wasn’t able to get it one single time. Any help appreciated. Thanks.</p>

**Comment by ghost** [comment]
<p>So I have copied from this one: <a href="https://github.community/t/github-package-registry-with-maven-causes-deployment-issues-error-code-400/17066">https://github.community/t5/GitHub-Actions/GitHub-Package-Registry-with-Maven-causes-deployment-issues/m-p/39879</a></p>
<pre><code># This workflow will build a Java project with Maven
# For more information see: https://help.github.com/actions/language-and-framework-guides/building-and-testing-java-with-maven

name: Java CI with Maven

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up JDK 13
      uses: actions/setup-java@v1
      with:
        java-version: 13
    - name: maven-settings
      uses: s4u/maven-settings-action@v2
      with:
        servers: '[{"id": "mixer", "username": "cryptearth", "password": "${GITHUB_TOKEN}"}]'
    - name: Build with Maven
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      run: mvn -B package --file pom.xml
</code></pre>
<p>But isn’t there any other way than using some actiong someone provides for free - wich may could change at any time?</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/cryptearth">@cryptearth</a> You could add settings.xml in the root path of your repo, you could refer to this article: </p>
<p><a href="https://help.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-apache-maven-for-use-with-github-packages#authenticating-with-a-personal-access-token" rel="nofollow noopener">https://help.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-apache-maven-for-use-with-github-packages#authenticating-with-a-personal-access-token </a></p>
<p>You could use ${env.PASSWORD} in the settings.xml. </p>
<pre><code>&lt;servers&gt;
    &lt;server&gt;
      &lt;id&gt;github&lt;/id&gt;
      &lt;username&gt;zhu5201314&lt;/username&gt;
      &lt;password&gt;${env.PASSWORD}&lt;/password&gt;
    &lt;/server&gt;
&lt;/servers&gt;
</code></pre>
<p>And in your workflow, add environment variable PASSWORD with the value of your personal access token stored in secrets. And add --settings parameter for <strong>mvn</strong> command. </p>
<pre><code>- name: Build with Maven
        env:
          PASSWORD: ${{ secrets.PackagePAT }}
        run: mvn -B package --file pom.xml --settings settings.xml
</code></pre>
<p>I noticed that you are using GITHUB_TOKEN for the password, in the official document, it is asked to use a PAT. </p>

**Comment by ghost** [comment]
<p><a class="mention" href="/yanjingzhu">@yanjingzhu</a> thank you for your reply.</p>
<blockquote>
<p>I noticed that you are using GITHUB_TOKEN for the password, in the official document, it is asked to use a PAT.</p>
</blockquote>
<p>Do you have any sources from where you got that? As the doc you linked only links to another page only explaining GITHUB_TOKEN. I didn’t read anything about </p>
<pre><code>secrets.PackagePAT
</code></pre>
<p>Also: Although adding a settings.xml to the repo and specify it in the mvn command in the workflow may is a solution it would also mean that in each clone and fork of the repo it get copied around as well wich may could end up my username spreading around with me losing controll over it. I’d rather not want this to happen. Is there a way to mark a file to not get cloned/forked? In addition to this it will only work in the workflow. So if someone clones it and uses the --settings switch local for what ever reason it would break. I’d prefer to not break other peoples stuff.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p><a class="mention" href="/cryptearth">@cryptearth</a> Sorry for the mistake I’ve been made. Yes , you could use GITHUB_TOKEN in GitHub Actions workflow.  </p>
<p>For the username issue, it doesn’t matter what’s the value it is. You could use github.actor context . </p>
<pre><code>- name: Build with Maven
   env:
      USENAME: ${{github.actor}}
      PASSWORD: ${{ secrets.GITHUB_TOKEN }}
   run: mvn -B package --file pom.xml --settings settings.xml
</code></pre>
<p>Then change the settings.xml &lt;servers&gt; section to :</p>
<pre><code>&lt;servers&gt;
    &lt;server&gt;
      &lt;id&gt;github&lt;/id&gt;
      &lt;username&gt;${env.USERNAME}&lt;/username&gt;
      &lt;password&gt;${env.PASSWORD}&lt;/password&gt;
    &lt;/server&gt;
  &lt;/servers&gt;
</code></pre>
<p>There is not a way to restrict a file to be cloned/forked. And, you are right, this could only works in workflow. </p>
<p>If someone clones the repo to their local side and try to build the maven package in local, they need to follow the <a href="https://help.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-apache-maven-for-use-with-github-packages#authenticating-to-github-packages" rel="nofollow noopener">doc</a> I linked before to change their local settings.xml with their personal access token.  </p>

**Comment by ghost** [comment]
<p><a class="mention" href="/yanjingzhu">@yanjingzhu</a> Thank you very much. I just implemented what you recommended and it works fine.</p>
<p>NetBeans seem to have some shadowing issues as it blends in the ~/.m2/settings.xml instead of the project local settings.xml - but that doesn’t cause any issues with the git commit and push. The workflow goes green. This should be somewhere noted in the official doc.</p>
